types/types.go

```diff
@@ -498,6 +516,8 @@ type SystemContext struct {
 	DockerInsecureSkipTLSVerify OptionalBool
 	// if nil, the library tries to parse ~/.docker/config.json to retrieve credentials
 	DockerAuthConfig *DockerAuthConfig
+	// if not nil, CryptoConfig will be used to encrypt/decrypt images
+	CryptoConfig *encconfig.CryptoConfig
```

Comment:
Is this really appropriate for `SystemContext`, or just for `Options`?

Also see the “OCI” naming speculation elsewhere.

Reply:
I think the thought here is that like `ImagePullSecrets`, `ImageDecryptSecrets` would be represented in a similar way. We looked at integrating this into `cri-o`, and using systemContext made things a lot cleaner - so as not to need to add the parameter too every struct which the flow of a request made its way to.

Reply:
The `SystemContext` …unwieldy type… exists primarily to insulate generic consumers of `types.ImageReference` from transport-specific parameters (so that `copy.Image` does not need to have explicit knowledge of “`docker://` transport parameters”, “`containers-storage:` parameters, “`dir:` parameters” and all the individual details directly in its API.

(Of course the transport dependency does not quite _go away_, it is here in `SystemContext`, and `copy.Image` needs to do fair amount of logic that is parametrized e.g. by the various `bool`-returning methods in `ImageDestination`, formulated as a generic facility, but transport-specific in intent/origin / choice of _those_ parameters and not others. But still, the `copy.Image` code is completely unaware of things like `DockerAuthConfig`, and in that sense `SystemContext` does its job well.)

Still, the way it collects _completely unrelated_ options, including options that could never actually be used at the same time, is unwieldy, impossible to generalize (e.g. to third-party transports, like some of the things Buildah does), and it makes option flow / usage difficult to reason about:

> using systemContext made things a lot cleaner - so as not to need to add the parameter too every struct which the flow of a request made its way to.

Another way to view the same thing is is that `SystemContext` made things a lot more confusing — the data was propagated everywhere without any ability to rely on the compiler to locate actual uses, or to report that a required value was not set by a caller.


Nothing in this PR adds any consumer of the encryption configuration in the transports or their helpers like `sysregistries2`, so this does not _need_ to be in `SystemContext` to preserve the current API. AFAICS the only consumer of the options is `copy.Image`, and `copy.Image` has `copy.Options`, a private, much better-scoped way to add options.

Reply:
Okay - yea this sounds good. I think I had some misconceptions about the role of systemContext. @harche and I had this discussion a while back as well, it seems clearer now that having an option is preferred.